src/tools/tools/process.fs

```diff
@@ -1171,10 +1171,146 @@ type internal ThreadPoolWait internal (id:int) as this =
+            let handleCollection = ThreadPoolWait.WaitHandleCollection.GetOrAdd( handle, fun _ -> ConcurrentDictionary())
+            handleCollection.Item(jobObject) <- ( rwh, infoFunc, ref DateTime.MinValue.Ticks )
+    static member private CallBack( state: Object ) (timeout:bool) = 
+        if not timeout then
+            try
+                let infoFunc,handle,continuation,unblockHandle,jobObject = state :?> ((unit->string)*WaitHandle*(unit->unit)*ManualResetEvent*Object)
+                let bExistHandleCollection, handleCollection = ThreadPoolWait.WaitHandleCollection.TryGetValue( handle )
+                let bExist, tuple =  
+                    if bExistHandleCollection then 
+                        handleCollection.TryGetValue(jobObject)
+                    else
+                        false, Unchecked.defaultof<_>
+                let ticksRef = 
+                    if not bExist then 
+                        ref DateTime.MinValue.Ticks // Continue to execute 
+                    else
+                        let _, _, ticksRef = tuple
+                        ticksRef
+                let bCancelled = ((!ticksRef) = DateTime.MaxValue.Ticks)
+                if not bCancelled then 
+                    ticksRef := DateTime.UtcNow.Ticks
+                    continuation() 
+                    if Utils.IsNotNull unblockHandle  then 
+                        // If there is an unblock handle, set it. 
+                        unblockHandle.Set() |> ignore  
+                if bExist then 
+                    let rwh, _, _ = tuple 
+                    rwh.Unregister(null) |> ignore 
+                    handleCollection.TryRemove( jobObject ) |> ignore 
+                    if handleCollection.IsEmpty then
+                        ThreadPoolWait.WaitHandleCollection.TryRemove( handle ) |> ignore 
+                    ThreadPoolWait.ToUnregister()
+                else
+                    rwhToRemove.Enqueue( handle, jobObject )
+            with 
+            | ex -> 
+                Logger.LogF( LogLevel.Error, fun _ -> sprintf "ThreadPoolWait.Callback, exception: %A" ex )
+                reraise()
+        else
+            failwith "ThreadPoolWait.Callback, timeout path reached (impossible execution route)"
+    static member ToUnregister() = 
+        let requeue = List<_>()        
+        while not rwhToRemove.IsEmpty do
+            let bRemove, removeTuple = rwhToRemove.TryDequeue()
+            if bRemove then 
+                let handle, jobObject = removeTuple
+                let bExistHandleCollection, handleCollection = ThreadPoolWait.WaitHandleCollection.TryGetValue( handle )
+                if bExistHandleCollection then 
+                    let bExist, tuple = handleCollection.TryGetValue(jobObject)  
+                    if bExist then 
+                        let rwh, _, _ = tuple       
+                        rwh.Unregister(null) |> ignore 
```

Comment:
On this code path, do you also need to remove the jobObject from the handleCollection, and remove the handleCollection from WaitHandleCollection?